src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs

```diff
@@ -173,6 +188,98 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object?> ar
+                    {
+                        pipe.Dispose();
+                    }
+                }
+
+                // Transfer ownership of the pipe to BufferedStream, it will dispose it:
+                using var stream = new BufferedStream(pipe, BufferSize);
+
+                using (var objectWriter = new ObjectWriter(stream, leaveOpen: true, cancellationToken))
+                {
+                    await dataWriter(objectWriter, data, cancellationToken).ConfigureAwait(false);
+                }
+
+                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception) when (cancellationToken.IsCancellationRequested)
+            {
+                // The stream has closed before we had chance to check cancellation.
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+        }
+
+        private static async Task ConnectPipeAsync(NamedPipeClientStream pipe, CancellationToken cancellationToken)
+        {
+            const int ConnectWithoutTimeout = 1;
+            const int MaxRetryAttemptsForFileNotFoundException = 3;
+            const int ErrorSemTimeoutHResult = unchecked((int)0x80070079);
+            var connectRetryInterval = TimeSpan.FromMilliseconds(20);
+
+            var retryCount = 0;
+            while (true)
+            {
+                try
+                {
+                    // Try connecting without wait.
+                    // Connecting with anything else will consume CPU causing a spin wait.
+                    pipe.Connect(ConnectWithoutTimeout);
+                    return;
+                }
+                catch (ObjectDisposedException)
+                {
+                    // Prefer to throw OperationCanceledException if the caller requested cancellation.
+                    cancellationToken.ThrowIfCancellationRequested();
+                    throw;
+                }
+                catch (IOException ex) when (ex.HResult == ErrorSemTimeoutHResult)
+                {
+                    // Ignore and retry.
+                }
+                catch (TimeoutException)
+                {
+                    // Ignore and retry.
+                }
+                catch (FileNotFoundException) when (retryCount < MaxRetryAttemptsForFileNotFoundException)
+                {
+                    // Ignore and retry
+                    retryCount++;
+                }
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await Task.Delay(connectRetryInterval, cancellationToken).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // To be consistent as to what type of exception is thrown when cancellation is requested,
+                    // always throw OperationCanceledException.
+                    cancellationToken.ThrowIfCancellationRequested();
```

Comment:
i don't really understand this comment.  isn't TaskCanceledException already a subclass of OperationCanceledException?  Do we do this anywhere else?  seems super weird.

Reply:
It is. IDK why is this here. Removed.